lib/SIL/Utils/OwnershipUtils.cpp

```diff
@@ -842,6 +859,85 @@ bool BorrowedValue::visitTransitiveLifetimeEndingUses(
   return true;
 }
 
+bool BorrowedValue::visitTransitiveGuaranteedForwardingUses(
+    function_ref<bool(Operand *)> visitor) const {
+  SmallVector<Operand *> guaranteedForwardingOps;
+
+  for (auto *use : value->getUses()) {
+    if (use->getOperandOwnership() != OperandOwnership::GuaranteedForwarding) {
+      continue;
+    }
+    if (!visitor(use)) {
+      return false;
+    }
+    guaranteedForwardingOps.push_back(use);
+  }
+
+  for (unsigned i = 0; i < guaranteedForwardingOps.size(); i++) {
+    for (auto val : guaranteedForwardingOps[i]->getUser()->getResults()) {
+      for (auto *valUse : val->getUses()) {
+        if (valUse->getOperandOwnership() !=
+            OperandOwnership::GuaranteedForwarding) {
+          continue;
+        }
+        if (!visitor(valUse)) {
+          return false;
+        }
+        guaranteedForwardingOps.push_back(valUse);
+      }
+    }
+  }
+
+  return true;
+}
+
+bool BorrowedValue::visitTransitiveGuaranteedForwardingPhiUses(
```

Comment:
I wasn't accurate when I said `visitTransitiveGuaranteedForwardingPhis` these will only be called on borrow value introducers. I have moved it out of `BorrowedValue` and added assert on values it can accept.

Reply:
Naming: `visitTransitiveGuaranteedForwardingPhis` maybe something like
`visitGuaranteedForwardingPhisWithinSSAScope`

Reply:
Fixed.